training/src/app/modules/chuck-norris/components/search/search.component.ts

```diff
@@ -0,0 +1,78 @@
+import { Component, Output, EventEmitter, OnInit } from '@angular/core';
+import { AbstractControl, FormBuilder, FormGroup, Validators } from "@angular/forms";
+import { debounceTime, Subject, takeUntil, tap } from "rxjs";
+
+
+@Component({
+	selector: 'app-search',
+	templateUrl: './search.component.html',
+	styleUrls: ['./search.component.scss'],
+
+})
+export class SearchComponent implements OnInit {
+
+	public searchForm: FormGroup;
+	private componentDestroyed$: Subject<void>                        = new Subject();
+	@Output() private generateJokesOnClickMode: EventEmitter<string>  = new EventEmitter<string>();
+	@Output() private generateJokesOnSearchMode: EventEmitter<string> = new EventEmitter<string>();
+	@Output() private buttonShouldByDisable: EventEmitter<boolean>    = new EventEmitter<boolean>();
+
+	constructor(private fb: FormBuilder) {
+	}
+
+	ngOnInit(): void {
+		this.searchForm = this.fb.group({
+			searchKey: [
+				'', [
+					Validators.required,
+					Validators.minLength(3),
+					Validators.maxLength(120)
+				]
+			],
+		});
+		this.getJokesWhenSearchKeyChange();
+	}
+
+	get searchKeyFormControl(): AbstractControl | null {
+		return this.searchForm.get('searchKey');
+	}
+
+	getJokesWhenSearchKeyChange(): void {
+		this.searchKeyFormControl?.valueChanges.pipe(
+			tap((searchValue: string) => {
+					if (this.searchKeyFormControl?.valid) {
+						this.generateJokesOnClickMode.emit(searchValue);
+						this.buttonShouldByDisable.emit(false);
+					} else {
+						this.buttonShouldByDisable.emit(true);
+					}
+				}
+			),
+			debounceTime(1000),
+			takeUntil(this.componentDestroyed$)
+		).subscribe((searchValue: string) => {
+			if (this.searchKeyFormControl?.valid) {
```

Comment:
you have the same logic in tap, so u are doing this twice. think if u can join them to a 1 place, or at least most of the code